crates/metadata-store/src/kv_memory_storage.rs

```diff
@@ -0,0 +1,180 @@
+                let result = self.delete(key, precondition);
+                if let Some(callback) = self.callbacks.remove(&request.request_id) {
+                    callback.complete_delete(result.map_err(Into::into));
+                }
+            }
+        }
+    }
+
+    pub fn get(&self, key: ByteString) -> Option<VersionedValue> {
+        self.kv_entries.get(&key).cloned()
+    }
+
+    pub fn get_version(&self, key: ByteString) -> Option<Version> {
+        self.kv_entries.get(&key).map(|entry| entry.version)
+    }
+
+    pub fn put(
+        &mut self,
+        key: ByteString,
+        value: VersionedValue,
+        precondition: Precondition,
+    ) -> Result<(), PreconditionViolation> {
+        match precondition {
+            Precondition::None => {
+                self.kv_entries.insert(key.clone(), value);
+            }
+            Precondition::DoesNotExist => {
+                if self.kv_entries.contains_key(&key) {
+                    return Err(PreconditionViolation::kv_pair_exists());
+                }
+
+                self.kv_entries.insert(key.clone(), value);
+            }
+            Precondition::MatchesVersion(expected_version) => {
+                let actual_version = self.kv_entries.get(&key).map(|entry| entry.version);
+
+                if actual_version == Some(expected_version) {
+                    self.kv_entries.insert(key.clone(), value);
+                } else {
+                    return Err(PreconditionViolation::version_mismatch(
+                        expected_version,
+                        actual_version,
+                    ));
+                }
+            }
+        }
+
+        // Not really happy about making the `KvMemoryStorage` aware of the NodesConfiguration. I
+        // couldn't find a better way to let a restarting metadata store know about the latest
+        // addresses of its peers which it reads from the NodesConfiguration. An alternative could
+        // be to not support changing addresses. Changing addresses will also only be possible as
+        // long as we maintain a quorum of running nodes. Otherwise, the nodes might not find each
+        // other to form quorum.
+        if let Some(metadata_writer) = self.metadata_writer.as_mut() {
+            if key == NODES_CONFIG_KEY {
+                let mut data = self
+                    .kv_entries
+                    .get(&key)
+                    .expect("to be present")
+                    .value
+                    .as_ref();
+                match StorageCodec::decode::<NodesConfiguration, _>(&mut data) {
+                    Ok(nodes_configuration) => {
+                        metadata_writer.submit(Arc::new(nodes_configuration));
+                    }
+                    Err(err) => {
+                        debug!("Failed deserializing NodesConfiguration: {err}");
+                    }
+                }
+            }
+        }
```

Comment:
Maybe it is better to not rely on the `NodesConfiguration` to learn about the other nodes' addresses and instead require that their addresses don't change because in the general case this cannot work (stopping all nodes and restarting them with different addresses).